ddht/tools/lru_sql_dict.py

```diff
@@ -0,0 +1,269 @@
+import sqlite3
+from typing import (
+    AbstractSet,
+    Any,
+    Callable,
+    Iterator,
+    MutableMapping,
+    NamedTuple,
+    Tuple,
+    TypeVar,
+    ValuesView,
+)
+
+from eth_utils.toolz import first
+
+#
+# SQL Schema
+#
+# key (primary key): bytes
+# value: bytes
+# pref: bytes - None if node is head, else points to the previous node's key
+# nref: bytes - None if node is tail, else points to the next node's key
+
+
+CREATE_CACHE_QUERY = """
+    CREATE TABLE IF NOT EXISTS cache (
+        key BLOB NOT NULL PRIMARY KEY,
+        value BLOB NOT NULL,
+        pref BLOB UNIQUE,
+        nref BLOB UNIQUE
+    )
+"""
+
+TKey = TypeVar("TKey")
+TValue = TypeVar("TValue")
+
+
+class Node(NamedTuple):
+    key: bytes
+    value: bytes
+    pref: bytes
+    nref: bytes
+
+
+class LRUSQLDict(MutableMapping[TKey, TValue]):
+    """
+    SQLite3-backed dictionary that implements an LRU cache.
+    """
+
+    def __init__(
+        self,
+        conn: sqlite3.Connection,
+        key_encoder: Callable[[TKey], bytes],
+        key_decoder: Callable[[bytes], TKey],
+        value_encoder: Callable[[TValue], bytes],
+        value_decoder: Callable[[bytes], TValue],
+        cache_size: int = None,
+    ) -> None:
+        self._conn = conn
+        self.cache_size = cache_size
+        self.key_encoder = key_encoder
+        self.key_decoder = key_decoder
+        self.value_encoder = value_encoder
+        self.value_decoder = value_decoder
```

Comment:
*nitpick*  Maybe make these private since I don't think it's appropriate for an external consumer to access them.